keps/sig-api-machinery/20190820-master-service-of-type-externalname.md

```diff
@@ -0,0 +1,299 @@
+
+This KEP is about adding support in Kubernetes for the master service to be of type `ExternalName` in addition to `ClusterIP`. This allows for kube-apiservers exposed via external FQDN (fully qualified domain name) to be directly consumed by Pods running in the cluster via that FQDN.
+
+## Motivation
+
+When exposing the kube-apiserver in HA mode, typically, a Layer-4 load-balancer is used. On different public/private cloud-providers, the endpoint of that load-balancer is either an IPv4/IPv6 address or a hostname (AWS being most prominent example for such offering). When consuming the latter offering, kubernetes cluster operators cannot use the hostname directly and have to disable the master endpoint reconciler in kube-apiserver with `--endpoint-reconciler-type=none` and write [custom controllers](https://github.com/gardener/aws-lb-readvertiser) which convert the hostname to IP addresses.
+
+It might be more suitable to advertise kube-apiserver's endpoint as FQDN, such as `api.my-cluster.example.com` for disaster recovery / prevention - in case of accidental deletion of said load-balancer and/or it's associated IP, the only thing required for recovery is to reconfigure the NS records to point to the newly created load-balancer.
+
+Historically there are several issues related to this problem:
+
+- https://github.com/kubernetes/kubernetes/issues/34043
+- https://github.com/kubernetes/kubernetes/issues/47588
+- https://github.com/coreos/tectonic-installer/issues/384
+
+### Goals
+
+- Make kube-servers exposed and advertised as FQDN, first class citizen in Kubernetes.
+- Add an additional reconciler in kube-apiserver for master kubernetes service of type `ExternalName`.
+- Support in kubelet for injecting the required environment variables `KUBERNETES_SERVICE_HOST` and `KUBERNETES_SERVICE_PORT` for [in-cluster discovery](https://github.com/kubernetes/client-go/blob/c8dc69f8a8bf8d8640493ce26688b26c7bfde8e6/rest/config.go#L399-L411) for master kubernetes service of type `ExternalName`.
+
+### Non-Goals
+
+- Environment variable support in kubelet for services of type `ExternalName` other than the master service. Related [comment](https://github.com/kubernetes/kubernetes/issues/60535#issuecomment-490757630).
+- In-cluster API server detection/auto-configuration other than environment variables named `KUBERNETES_SERVICE_HOST` and `KUBERNETES_SERVICE_PORT`.
+- Network connectivity from Nodes/Pods to the kube-server is out of scope and is assumed to be properly configured and working. Pods using the hostnetwork or CNI network are able to initiate connections to kube-apiserver.
+
+## Proposal
+
+In order to support this feature in Kubernetes the following changes are proposed:
+
+### Kube-apiserver
+
+An additional reconciler to be introduced in kube-apiserver.
```

Comment:
Can we avoid that? the current can be modified to support both modes with feature gates.

Reply:
It could be done. The "main" [logic flow](https://github.com/kubernetes/kubernetes/blob/93c06821e623fdfb5fd1988f850c3c35bcaf2a5f/pkg/master/controller.go#L239-L247)  is simple enough to support this.

I wrote this comment mainly because of this [comment](https://github.com/kubernetes/kubernetes/blob/93c06821e623fdfb5fd1988f850c3c35bcaf2a5f/pkg/master/controller.go#L231-L234). While I wait for the KEP to be merged I'll refactor the controller to use watches and not do polling.